### ISSUE TITLE
Migrate from Bors to GitHub merge queues

### ISSUE THREAD

**Issue by justinvp** [pulumi/pulumi#13501]
The community instance of Bors has been deprecated. We need to move off Bors to GitHub merge queues.

**Comment by joeduffy** [comment]
Can someone summarize the pros and cons of using a merge queue? Full transparency, although I have managed teams with merge queues many times, sub-50 developers it's almost always more hassle than it's worth. As a casual observer, I would have said bors has been a similar experience to this. But perhaps if I saw the full analysis written down, I'd think differently.

**Comment by blampe** [comment]
@joeduffy FWIW I started with Bors and migrated to GHMQ at my last startup. The GHMQ UX is much nicer (as you would expect), and if you've setup Bors all the hard work is already done (or it was in my case -- I just needed to change some branch names in my CI jobs). For the most part GHMQ just re-uses your PR test suite, so it's not a maintenance burden. I wouldn't worry about the effort involved, if that's part of your concern. (Unless this CI setup is radically different from what I'm imagining -- I don't have all the context!)

Re: why do you need it at all, I'll share that I found it _enormously_ helpful to include a `pulumi up` as part of our merge queue's test suite. Previews on PRs are _OK_ but it's still possible to land code that doesn't deploy. Instead, we were able to deploy every commit to staging _before_ merging, which eliminated _so many_ failed updates we had been resolving manually. pu/pu's situation is different, of course, but life is much simpler when you can make guarantees about every SHA in `main`. There's probably something analogous wrt publishable binaries or similar.

**Comment by AaronFriel** [comment]
Hey @joeduffy, I implemented bors on the pulumi/pulumi repository. I'm certain that removing merge queues will reintroduce the panoply of woes we addressed last year. It would be a significant regression to our release process and commitment to quality. 

I'll address the cons first: the largest is educating engineers in onboarding about merge queues. Really - I think that's it. 

> Full transparency, although I have managed teams with merge queues many times, sub-50 developers it's almost always more hassle than it's worth. 

What Uber's Engineering team evangelized in their paper [Keeping master green at scale](https://www.uber.com/blog/research/keeping-master-green-at-scale/) is now trickling down into products. GitLab ships merge queues as part of their core product (called merge trains), and for premium SKU users, it is the default "auto-merge" behavior. Other combined VCS & CI/CD systems are following suit, such as GitHub here.

## Background

In early 2022 when I onboarded, we spent an enormous amount of time firefighting CI/CD. Whether it was trunk failing tests, CI/CD outages, or long cycle times, productivity and morale in contributing to this repository were suffering. I proposed a design doc, [Always Green: Keeping Trunk Stable (Google Docs)](https://docs.google.com/document/d/1aMgzlybQEialalP3PQFLMOxC7aCLy4warA3TgC-Ns4Q/edit#heading=h.ypffrqjtglh1), and implemented it around this time a year ago using `bors`. As an aside: GitHub's implementation will be substantially nicer to use, and remove a point of failure on the `bors` service.

After implementing this design, [we dramatically improved our CI situation (slack)](https://pulumi.slack.com/archives/C02VCJEBT2N/p1674775742630089):

> Checking my notes:
> * Master being red went from over 1/3 of the time, blocking merges and releases and causing P0 incidents, to 0% of the time.
> * During Q3, CI cycle times fell from weekly averages of 124 minutes to 22 minutes per PR, and much less if you count the fact that bors sometimes merges 3 or 4 PRs at a time and consecutive PRs don't introduce risks of P0s.
> * Worst case CI times fell from >300 minutes to about 45 minutes and I don't have metrics, but I believe our stddev also fell significantly. (We have a tool that gets this data, but it's finnicky.)
90th percentile release time (aka "mean time to recovery") fell from >600 minutes to about 30 minutes.
> * We had several days in Q2 and Q3 where releases were blocked by failing master build and/or CI cycle times of multiple hours. The release process now takes about as long as the version update PR takes to merge - which we could probably optimize but haven't invested any effort in.

In another thread, I wrote:

> I find the value of merge queues isn't in scaling the number of engineers, but in **scaling testing** (we run over 12 hours of tests on every release now, while retaining the benefits below) **and release processes**, ensuring that master is always green.
> 
> In addition to keeping master always green, we reduced time to release master by >99%, time to recovery (merge a fix & release) by up to 90%, reduced overall GitHub CI usage on PRs to our largest repository on free GitHub runner usage (particularly acute - up to 6 hour queues before merge queues), while allowing merge queues pipelines to fan out, and we adopted a build, test, and release process where the released artifacts are (not "equivalent to", but are) those built and tested in the last merge. There is no daylight between what CI tests and what we release to customers.

**Comment by Frassle** [comment]
> I'll address the cons first: the largest is educating engineers in onboarding about merge queues. Really - I think that's it.

I'd disagree with that. There is one large negative we keep hitting with merge queue (at least with the bors implementation) which we didn't have when we just ran PR jobs, and that is that intermittent failures (Mostly infrastructure failures like setting up Go, but also just random unexplained crashes we see on Windows) require a full retry of all jobs.
When we had PR jobs if one chain failed because of a network blip or if Windows tests just flaked out as they do you could easily re-run _just that one_. The time difference of this is obviously massive.

That was the main reason I raised https://github.com/pulumi/pulumi/issues/11449, although there's a couple of other negatives listed in that issue as well.

On overall balance I feel we're slightly better off now than we were before, but I think a lot of that is from work that happened to get done at the same time as adding merge queues, like the changing of the changelog system (which caused a lot of rebase conflicts before, but doesn't need to be used with merge queues), and increasing test coverage to run integration tests before merging to master (which before often failed on master as they weren't run automatically for PRs).

**Comment by AaronFriel** [comment]
@Frassle raises a good point, and it can be a con, especially when introducing merge queues - we definitely faced those growing pains post launch. 

If tests are flaky, merge queues are a forcing function to address that, as otherwise you cannot deliver software. I'd argue this is a good thing, but I also believe failing tests should be fixed or deleted if they're not providing any value.

The ongoing Windows disconnection issues: https://github.com/pulumi/pulumi/actions/runs/5802184132/job/15728250681

Are a huge bummer. I wish we could fix this systematically. Will take that conversation async.

**Comment by Frassle** [comment]
> If tests are flaky, merge queues are a forcing function to address that, as otherwise you cannot deliver software. I'd argue this is a good thing, but I also believe failing tests should be fixed or deleted if they're not providing any value.

I'd agree that merge queues are good at forcing you to fix flaky tests, and I have no qualms with just _having to deal with that_. My main gripe is that infrastructure is flaky as well and there's less we can do about that. Network deciding to have a hicup, monitoring software failing, package listing missing a package due to a server glitch, etc. Those are primarily the sorts of failures I see day to day using bors, not our tests being intermittent (it does happen, but it's much much less common).

**Comment by justinvp** [comment]
I suggest we try switching to GHMQ and see how it is compared to bors, and then re-evaluate after a period of time using it. The effort to move to GHMQ should be minimal to Bryce's point. We've already done the hard work setting up CI for merge queues with bors. We just need to do some minor reconfiguration to move to GHMQ, starting with https://github.com/pulumi/pulumi/pull/13681.

Moving from bors to not using merge queues is likely a much more involved reworking of CI.
